src/analyses/ai_storage.h

```diff
@@ -0,0 +1,298 @@
+  typedef std::shared_ptr<const trace_sett> ctrace_set_ptrt;
+  typedef goto_programt::const_targett locationt;
+
+  /// Returns all of the histories that have reached
+  /// the start of the instruction.
+  virtual ctrace_set_ptrt abstract_traces_before(locationt l) const = 0;
+
+  /// Non-modifying access to the stored domains,
+  /// used in the ai_baset public interface.
+  /// In the case of un-analysed locals this should create a domain
+  /// The history version is the primary version, the location one may
+  /// simply join all of the histories that reach the given location
+  virtual cstate_ptrt abstract_state_before(
+    trace_ptrt p,
+    const ai_domain_factory_baset &fac) const = 0;
+
+  virtual cstate_ptrt abstract_state_before(
+    locationt l,
+    const ai_domain_factory_baset &fac) const = 0;
+
+  /// Look up the analysis state for a given history,
+  /// instantiating a new domain if required.
+  virtual statet &
+  get_state(trace_ptrt p, const ai_domain_factory_baset &fac) = 0;
+
+  /// Reset the abstract state
+  virtual void clear()
+  {
+    return;
+  }
+
+  /// Notifies the storage that the user will not need the domain object(s)
+  /// for this location.  After this has been called abstract_state_before may
+  /// return an over-approximation of the value and get_state may give an
+  /// under-approximation (forcing recomputation).
+  /// If there are multiple histories that reach this location all will be
+  /// affected
+  virtual void prune(locationt l)
+  {
+    return;
+  }
+};
+
+// There are a number of options for how to store the history objects.
+// This implements a simple one.  It is not in ai_storage_baset so that
+// storage implementations can implement their own, more specific, approaches
+class trace_map_storaget : public ai_storage_baset
+{
+protected:
+  typedef std::map<locationt, trace_set_ptrt> trace_mapt;
+  trace_mapt trace_map;
+
+  // This retains one part of a shared_ptr to the history object
+  void register_trace(trace_ptrt p)
+  {
+    // Save the trace_ptrt
+    trace_mapt::iterator it = trace_map.find(p->current_location());
+    if(it == trace_map.end())
+    {
+      trace_set_ptrt s(new trace_sett());
+      auto ins = trace_map.emplace(p->current_location(), s);
+      CHECK_RETURN(ins.second);
+      it = ins.first;
+    }
+    // Strictly this should be "it->second points to a trace_set"
+    POSTCONDITION(it->second != nullptr);
+
+    it->second->insert(p);
+
+    return;
+  }
+
+public:
+  ctrace_set_ptrt abstract_traces_before(locationt l) const override
+  {
+    trace_mapt::const_iterator it = trace_map.find(l);
+    if(it == trace_map.end())
+      return trace_set_ptrt(new trace_sett());
+
+    // Strictly this should be "it->second points to a trace_set"
+    POSTCONDITION(it->second != nullptr);
+    return it->second;
+  }
+
+  void clear() override
+  {
+    ai_storage_baset::clear();
+    trace_map.clear();
+    return;
```

Comment:
I tend to put explicit return statements into things with void return type so that it is clear that it is done, I haven't just got part way through and wandered off and forgotten what I was doing.  Not that that has ever happened.